src/atomEffect.ts

```diff
@@ -1,167 +1,61 @@
-    set(rerunAtom, toggle)
-  }
-
-  const rerunAtom = atom(true)
-  rerunAtom.debugLabel = 'rerunAtom'
-  if (process.env.NODE_ENV !== 'production') {
-    rerunAtom.debugPrivate = true
-  }
-
-  const baseAtom = atom<void, [WriteFn], PromiseOrValue<void>>(
-    async (get, { setSelf }) => {
-      const internalState = get(internalStateAtom)
-      ensureWatchDependencies(get) // ensure all watched atoms are captured by Jotai's internal dependency map
-      if (!internalState) return
-      if (internalState.inProgress) return // effect setter has triggered this run
-      if (!isFirstRun(get) && !hasChanged(get)) return // no changes
-      internalState?.dependencyMap.clear()
-      const shouldCollectDeps = { value: true }
-      const getter = makeGetter(get, shouldCollectDeps)
-      getter(rerunAtom)
-      await defer() // defer effectFn to allow other atoms to update synchronously first
-      setSelf(async (_, set) => {
-        const setter = makeSetter(set, () => get(internalStateAtom))
-        await internalState.cleanup?.()
-        internalState.cleanup = await effectFn(getter, setter)
-        shouldCollectDeps.value = false
-        rerunToEvaluate(setter) // rerun to ensure all dependencies are captured
-      })
-    },
-    (get, set, writeFn) => writeFn(get, set)
-  )
-
-  if (process.env.NODE_ENV !== 'production') {
-    baseAtom.debugPrivate = true
-  }
-
-  const effectAtom = atom((get) => {
-    get(baseAtom)
-  })
-  return effectAtom
-}
-
-export function makeInternalStateAtom(
-  internalStateFactory: () => InternalState = createInternalState
-) {
-  const internalStateAtom = atom<
-    InternalState | null,
-    [boolean | InternalState | null],
-    void
-  >(null, (get, set, isInit) => {
-    if (isInit) {
-      set(internalStateAtom, internalStateFactory())
+  const refAtom = atom(() => ({
+    mounted: false,
+    inProgress: false,
+    cleanup: undefined as void | CleanupFn,
+  }))
+
+  const refreshAtom = atom(0)
+
+  const initAtom = atom(null, (get, set, mounted: boolean) => {
+    const ref = get(refAtom)
+    if (mounted) {
+      ref.mounted = true
+      set(refreshAtom, (c) => c + 1)
     } else {
-      ;(async () => get(internalStateAtom)?.cleanup?.())()
-      set(internalStateAtom, null)
+      ref.cleanup?.() // do not await
+      ref.cleanup = undefined
+      ref.mounted = false
     }
   })
-  internalStateAtom.onMount = (init) => {
+  initAtom.onMount = (init) => {
     init(true)
     return () => init(false)
   }
-  internalStateAtom.debugLabel = 'internalStateAtom'
 
-  if (process.env.NODE_ENV !== 'production') {
-    internalStateAtom.debugPrivate = true
-  }
-
-  return internalStateAtom
-}
+  const effectAtom = atom(
+    async (get, { setSelf }) => {
+      get(refreshAtom)
+      const ref = get(refAtom)
+      if (!ref.mounted) {
+        return
+      }
+      if (ref.inProgress) {
+        throw new Error('infinite loop detected')
+      }
+      ref.inProgress = true
```

Comment:
effectAtoms can stack if effectFn suspends while a dependency update triggers another rerun. 
I recommend using a counter instead of booleans here.

Reply:
Also, I still think inProgress should wrap setSelf otherwise async calls to the setter would cause an infinite loop as inProgress would be set to false after effectFn promise resolves.

Reply:
I haven't thought about such a case. Sure, I can use a counter.